design/observerability-2.md

```diff
@@ -0,0 +1,196 @@
+
+NOTE: The diagram implementation and specific charts might change during the implementation phase. The above proposal
+shows only the general idea.
+
+The proposal consists of 3 sections:
+- System health - Enables the SRE Team to quickly tell if Keycloak cluster is up and running. This section consists of:
+  - Readiness and Liveness Probes chart
+  - HTTP 5xx error rate for all major Keycloak Services, including: Admin Console, Account Console, OIDC, SAML
+  - Aggregated latency across the most important endpoints
+  - (Optional) Error rate in logs [^1]
+- Detailed Metrics - This section helps foresee upcoming failures. It includes:
+  - Pod memory, CPU and container storage charts
+  - GC Times
+  - Heap and non-heap memory metrics
+- Experimental metrics - Acts as a bridge between the SRE and Keycloak Teams. We introduce new metrics here and if they
+are useful, we promote them to other sections. This section includes:
+  - External IdP state (Social Login Providers)
+  - The Total number of CRs discovered by the Operator vs the number of CRs successfully reconciled
+
+[^1]: The Error rate in logs is not a required metric for the dashboard. After discussing it with both AOS and MAS
+Teams we came to the conclusion that it might be fuzzy and inaccurate. A much better choice is to observe responses
+from a service and use logs for further, more detailed diagnosis.
+
+## SLI/SLO/SLA
+
+SLA might specify SLOs in several categories, including:
+- Availability
+- Latency
+- Freshness - cache usage
+- Correctness - interesting in distributed environments
+- Completeness - interesting for batch applications
+
+Based on the dashboard, the proposed SLI/SLO and SLA are the following:
+
+|   Use Case             | Type         | SLI                                                                           | SLO           | SRE Team Action if failing                               |
+|------------------------|--------------|-------------------------------------------------------------------------------|---------------|----------------------------------------------------------|
+| Health                 | Availability | Check individual Pod Liveness and Readiness status                            | To be defined | Check the DB status  / Report a JIRA / Restart Pods      |
+| Health                 | Error rate   | Monitor error rates of individual Keycloak services                           | To be defined | Investigate further                                      |
+| Health (optional)      | Error rate   | Monitor error rate in logs                                                    | To be defined | If increases, check what might be causing it             |
+| Performance            | Latency      | Monitor the aggregated average response time for individual Keycloak Services | To be defined | If drastically increases, check what might be causing it |
+
+For all the SLOs mentioned the above, the Keycloak Operator needs to set proper Prometheus Alerts.
```

Comment:
This is useful.
Is there a plan to document the `SRE Team Action if failing` in more detail? i.e a SOP or runbook

Reply:
I think such a document should be written as part of the documentation. Let me mention this explicitly.